### PR TITLE
[WIP] Migrate DynamicGroup model and functionality from extras to core

### PR DESCRIPTION
**Pull request by glennmatthews**
# Closes: #2826 
# What's Changed

- Move the `DynamicGroup` model and associated classes (filters, forms, views, API, etc.) from the `extras` app to the `core` app.
- Create migrations (based on example from https://davit.tech/django-move-model) to move the existing database tables and content-types between apps instead of deleting the existing tables and creating new empty tables.
    - Migrating the content-type isn't covered in the linked blog post but it's necessary to ensure that GFKs such as in ObjectChange don't get broken.
    - I have manually tested migration back-and-forth between this branch and next with a DynamicGroup with a change history; I haven't tested the migrations against an empty table but CI should prove that out one way or the other.
- To get URL patterns to work correctly, I had to split `nautobot.core.urls` into `nautobot.core.urls` (app-level URLs, such as now dynamic-groups) and `nautobot.core.urls_root` (the base urlconf for all of Nautobot), and similarly split `nautobot.core.api.urls` into `nautobot.core.api.urls` and `nautobot.core.api.urls_root`.
- To avoid circular imports between `extras` and `core` after the move, I had to:
    - convert `nautobot.core.api.views` and `nautobot.core.views` into submodules
    - move the setup of the dynamic-group signal handlers into the AppConfig's `ready()` method instead of connecting them at import time
- One lingering issue I haven't been able to solve for is adding any nav menu items in `nautobot.core.navigation`. Some issue with application loading and initialization order - the URL resolution for the nav menu items causes `nautobot.core.admin` to be imported and evaluated early, apparently too early? and so app startup fails.

# TODO
- [x] Explanation of Change(s)
- n/a Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- n/a Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [x] Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/core/api/__init__.py [hunk 1]
```diff
@@ -5,11 +5,11 @@
     TimeZoneSerializerField,
 )
 from .routers import OrderedDefaultRouter
+from .nested_serializers import WritableNestedSerializer
 from .serializers import (
     BaseModelSerializer,
     BulkOperationSerializer,
     ValidatedModelSerializer,
-    WritableNestedSerializer,
```
Comment (glennmatthews) [inline nautobot/core/api/__init__.py:12]: Disregard changes to this file - artifact of some early attempts at refactoring that turned out to be unnecessary.

File: nautobot/core/api/serializers/__init__.py [hunk 2]
```diff
@@ -139,37 +139,6 @@ def get_field_names(self, declared_fields, info):
         return fields
 
 
-class ValidatedModelSerializer(BaseModelSerializer):
```
Comment (glennmatthews) [inline nautobot/core/api/serializers/__init__.py:142]: Disregard changes to this file (other than converting it into a submodule) - the relocation of `ValidatedModelSerializer` within this file are leftovers from some early refactoring that proved unnecessary.

File: nautobot/core/navigation.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+from nautobot.core.apps import NavMenuAddButton, NavMenuGroup, NavMenuItem, NavMenuTab
+
+
+# TODO fix initialization order issue with admin site
+_menu_items = (
```
Comment (glennmatthews) [inline nautobot/core/navigation.py:5]: As noted in the PR description, setting `menu_items` here causes an order-of-initialization problem with `nautobot.core.admin`. TODO.

File: nautobot/core/api/urls.py [hunk 4]
```diff
@@ -1,57 +1,29 @@
...
-    StatusView,
-    NautobotSpectacularSwaggerView,
-    NautobotSpectacularRedocView,
-)
-from nautobot.extras.plugins.urls import plugin_api_patterns
+from nautobot.core.api.views.dynamic_groups import DynamicGroupViewSet, DynamicGroupMembershipViewSet
+from . import routers, views
 
+router = routers.OrderedDefaultRouter()
+# router.APIRootView = views.CoreRootView
```
Comment (glennmatthews) [inline nautobot/core/api/urls.py:7]: Uncommenting this line (following a similar pattern for other Nautobot apps' `api.urls` code) results in an error from inside Django. I haven't chased it down.

**Comment by glennmatthews** [comment]
OK, that's a bizarre error from unit test initialization. Any ideas? Will investigate on Monday.

**Comment by glennmatthews** [comment]
Seems to be an issue with the test-only model defined in [`nautobot/core/tests/test_models.py`](https://github.com/nautobot/nautobot/blob/develop/nautobot/core/tests/test_models.py) not getting a table created in the database. Not sure why my changes are influencing anything here; but based on some googling it perhaps shouldn't be working in the first place as written? Could use some extra eyes here.

**Comment by glennmatthews** [comment]
One extra concern not directly addressed by this PR is how this change would be handled by apps (plugins) that have foreign-keys to any of the Extras models, such as to in this case a DynamicGroup. Probably the data migration needs to inspect the migrated model for any and all reverse-relations that it contains, and handle all of those identified foreign-keys at once.

**Comment by bryanculver** [comment]
😢 

At the moment we're going to shelve moving extras for the time being (maybe 3.0) due to the significant impact to migration for 2.0. Let's let the Plugin API be more widely adopted and clean up some other things first before we move extras.

**Pull request closed without merging**